### PR TITLE
feat(tencent): support tencent provider

### PR DESCRIPTION
**Pull request by NewGr8Player**
relate to #54

### CODE REVIEW

File: docs/tencent/README.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+### Setup K3s Cluster
+If already have access information in `$HOME/.autok3s/config.yaml` you can use the simplified command.
+```bash
+autok3s create \
+    --provider tencent \
+    --region <region> \
+    --zone <zone> \
+    --name <cluster name> \
+    --security-group <security-group id> \
+    --vpc <vpc id> \
```
Comment (Jason-ZW) [inline docs/tencent/README.md:19]: Need to pass two arguments at the same time？ How about only pass `--subnet`?

File: docs/tencent/README.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    --subnet <subnet id> \
+    --key-pair <key-pair id> \
+    --ssh-user <ssh-user> \
```
Comment (Jason-ZW) [inline docs/tencent/README.md:22]: default `ssh-user`, not workers?

File: docs/tencent/README.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+### Join K3s Nodes
+If you have ever created a cluster using `autok3s` on your current machine, you can use the simplified command.
+```bash
+autok3s join \
+    --provider tencent \
+    --region <region> \
+    --zone <zone> \
+    --name <cluster name> \
+    --ssh-user <ssh-user> \
```
Comment (Jason-ZW) [inline docs/tencent/README.md:72]: same question

File: docs/tencent/ram.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# RAMs
+Tencent Cloud Provider RAMs:
+```json
+{
+    "version": "2.0",
+    "statement": [
+        {
+            "action": [
```
Comment (Jason-ZW) [inline docs/tencent/ram.md:8]: Does this include THE RAMS of CCM?

File: pkg/hosts/dialer.go [hunk 5]
```diff
@@ -52,7 +52,7 @@ func (d *Dialer) OpenTunnel(timeout bool) (*Tunnel, error) {
 	if err := wait.ExponentialBackoff(common.Backoff, func() (bool, error) {
 		conn, err = d.getSSHTunnelConnection(timeout)
 		if err != nil {
-			return false, err
+			return false, nil
```
Comment (Jason-ZW) [inline pkg/hosts/dialer.go:55]: why return nil?

File: pkg/providers/tencent/flag.go [hunk 6]
```diff
@@ -0,0 +1,775 @@
+package tencent
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/cnrancher/autok3s/pkg/cluster"
```
Comment (Jason-ZW) [inline pkg/providers/tencent/flag.go:8]: please format package import order

File: pkg/providers/tencent/tencent.go [hunk 7]
```diff
@@ -0,0 +1,1431 @@
...
+					tencentCCM.VpcID, tencentCCM.NetworkRouteTableName, defaultCidr)
+			} else {
+				tmpl = fmt.Sprintf(tencentCCMTmpl, tencentCCM.Region, tencentCCM.SecretID, tencentCCM.SecretKey,
+					tencentCCM.VpcID, tencentCCM.NetworkRouteTableName, p.ClusterCIDR)
+			}
+
+			extraManifests = append(extraManifests, fmt.Sprintf(deployCCMCommand,
+				base64.StdEncoding.EncodeToString([]byte(tmpl)), common.K3sManifestsDir))
+		}
+		p.logger.Infof("[%s] start deploy Tencent additional manifests\n", p.GetProviderName())
```
Comment (Jason-ZW) [inline pkg/providers/tencent/tencent.go:168]: lowercase Tencent

File: pkg/providers/tencent/tencent.go [hunk 8]
```diff
@@ -0,0 +1,1431 @@
...
+		if err := cluster.DeployExtraManifest(c, extraManifests); err != nil {
+			if c.CloudControllerManager {
+				if _, err := p.deleteClusterRouteTable(); err != nil {
+					p.logger.Errorf("[%s] error call deleteClusterRouteTable, msg: %v", p.GetProviderName(), err)
+				}
+			}
+			return err
+		}
+		p.logger.Infof("[%s] successfully deploy Tencent additional manifests\n", p.GetProviderName())
```
Comment (Jason-ZW) [inline pkg/providers/tencent/tencent.go:177]: lowercase Tencent

File: pkg/providers/tencent/tencent.go [hunk 9]
```diff
@@ -0,0 +1,1431 @@
...
+	}
+	return true, ids, err
+}
+
+func (p *Tencent) GenerateMasterExtraArgs(cluster *types.Cluster, master types.Node) string {
+	if option, ok := cluster.Options.(tencent.Options); ok {
+		if cluster.CloudControllerManager {
+			extraArgs := fmt.Sprintf(" --kubelet-arg=provider-id=%s.%s --node-name=%s.%s",
+				option.Region, master.InstanceID, option.Region, master.InstanceID)
+			return extraArgs
```
Comment (Jason-ZW) [inline pkg/providers/tencent/tencent.go:435]: Does this meet The requirements of Tencent CCM?

**Review by Jason-ZW — CHANGES_REQUESTED** [review]
Others LGTM

File: pkg/providers/tencent/tencent.go [hunk 10]
```diff
@@ -0,0 +1,1431 @@
...
+		}
+	}
+	return true, ids, err
+}
+
+func (p *Tencent) GenerateMasterExtraArgs(cluster *types.Cluster, master types.Node) string {
+	if option, ok := cluster.Options.(tencent.Options); ok {
+		if cluster.CloudControllerManager {
+			extraArgs := fmt.Sprintf(" --kubelet-arg=provider-id=%s.%s,--cloud-provider=external,--node-status-update-frequency=30s --node-name=%s",
+				option.Region, master.InstanceID, master.InternalIPAddress[0])
```
Comment (NewGr8Player) [inline pkg/providers/tencent/tencent.go:434]: add `--cloud-provider=external,--node-status-update-frequency=30s` for `kubelet`
change nodename to node private ip

**Pull request merged**